File: ServiceCore/ActivityService.cs
using BOCore;
using FacadeCore;
using DALCore;
using DALCore.Models;
using ServiceCore.Translators;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;

namespace ServiceCore
{
    public class ActivityService : IActivityService
    {
        private readonly UnitOfWorkCore _uow;

        public ActivityService(UnitOfWorkCore uow)
        {
            _uow = uow ?? throw new ArgumentNullException(nameof(uow));
        }

        public GetResponse<IdNameBO> GetActivitiesForSelect()
        {
            var response = new GetResponse<IdNameBO>();

            var entities = _uow.Activities
                .GetNoTracking()
                .Include(m => m.Group)
                .OrderBy(m => m.Omschrijving)
                .ToList();

            foreach (var e in entities)
                response.AddValue(e.GetIdName());

            // Sorteren op Display (de OrderBy bovenaan sorteert op Omschrijving in DB)
            // Hier nog eens voor de zekerheid op de uiteindelijke Display.
            if (response.Values != null && response.Values.Count > 1)
                response.Values = response.Values.OrderBy(v => v.Display).ToList();

            return response;
        }

        public GetResponse<ActivityBO> GetActivities()
        {
            var response = new GetResponse<ActivityBO>();

            var entities = _uow.Activities
                .GetNoTracking()
                .Include(m => m.Group)
                .OrderBy(m => m.Omschrijving)
                .ToList();

            foreach (var e in entities)
            {
                var bo = new ActivityBO
                {
                    ID = e.ActivityId,
                    Name = e.Omschrijving
                };

                if (e.Group != null)
                {
                    bo.Group.Name = e.Group.Name;
                    bo.Group.ID = e.Group.GroupId;
                    bo.Group.Lot = Convert.ToInt16(e.Group.Lot);
                }

                response.AddValue(bo);
            }

            return response;
        }

        public GetResponse<ActivityBO> GetActivitiesbyId(List<int> idList)
        {
            var response = new GetResponse<ActivityBO>();
            if (idList == null || idList.Count == 0) return response;

            var entities = _uow.Activities
                .GetNoTracking()
                .Include(m => m.Group)
                .Where(m => idList.Contains(m.ActivityId))
                .ToList();

            foreach (var e in entities)
            {
                var bo = new ActivityBO
                {
                    ID = e.ActivityId,
                    Name = e.Omschrijving
                };

                if (e.Group != null)
                {
                    bo.Group.Name = e.Group.Name;
                    bo.Group.ID = e.Group.GroupId;
                    bo.Group.Lot = Convert.ToInt16(e.Group.Lot);
                }

                response.AddValue(bo);
            }

            return response;
        }

        public GetResponse<ActivityBO> GetActivitybyId(int id)
        {
            var response = new GetResponse<ActivityBO>();

            var entity = _uow.Activities
                .GetNoTracking()
                .Include(m => m.Group)
                .SingleOrDefault(m => m.ActivityId == id);

            if (entity != null)
            {
                var bo = new ActivityBO
                {
                    ID = entity.ActivityId,
                    Name = entity.Omschrijving,
                    Group = entity.Group != null
                        ? new ActivityGroupBO
                        {
                            ID = entity.Group.GroupId,
                            Name = entity.Group.Name,
                            Lot = Convert.ToInt16(entity.Group.Lot)
                        }
                        : null
                };

                response.AddValue(bo);
            }

            return response;
        }
        public GetResponse<ActivityGroupBO> GetActivityGroups()
        {
            var response = new GetResponse<ActivityGroupBO>();

            var entities = _uow.ActivityGroups.GetNoTracking()
                .Include(g => g.Activity)               // we hebben de activiteiten vaak nodig
                .OrderBy(g => g.Lot).ThenBy(g => g.Name);

            foreach (var entity in entities)
            {
                var bo = new ActivityGroupBO();
                var err = ActivityGroupTranslator.TranslateEntityToBO(entity, bo);
                if (err == ErrorCode.Success)
                    response.AddValue(bo);
                else
                    response.AddError(err.ToString());
            }

            return response;
        }

        public GetResponse<IdNameBO> GetActivityGroupsForSelect()
        {
            var response = new GetResponse<IdNameBO>();

            var entities = _uow.ActivityGroups.GetNoTracking()
                .OrderBy(g => g.Lot).ThenBy(g => g.Name)
                .ToList();

            foreach (var entity in entities)
                response.AddValue(entity.GetIdName());

            return response;
        }

        public GetResponse<ActivityGroupBO> GetActivityGroupbyId(int id)
        {
            var response = new GetResponse<ActivityGroupBO>();

            // Belangrijk: GetById() gebruikt Find() en laadt navigaties niet.
            var entity = _uow.ActivityGroups.GetNoTracking()
                .Include(g => g.Activity)
                .SingleOrDefault(g => g.GroupId == id);

            if (entity == null)
            {
                response.AddError("activity group not found");
                return response;
            }

            var bo = new ActivityGroupBO();
            var err = ActivityGroupTranslator.TranslateEntityToBO(entity, bo);
            if (err == ErrorCode.Success)
                response.AddValue(bo);
            else
                response.AddError(err.ToString());

            return response;
        }
        public Response InsertUpdate(ActivityBO bo)
        {
            var response = new Response();

            if (string.IsNullOrWhiteSpace(bo?.Name))
            {
                response.AddError("name is mandatory");
                return response;
            }

            Activity entity = null;

            if (bo.ID == 0)
                entity = _uow.Activities.GetNew();
            else
                entity = _uow.Activities.GetById(bo.ID);

            if (entity == null)
            {
                response.AddError("activity not found");
                return response;
            }

            var err = ActivityTranslator.TranslateBOToEntity(entity, bo);
            if (err != ErrorCode.Success)
            {
                response.AddError(err.ToString());
                return response;
            }

            var result = _uow.SaveChangesAsync().GetAwaiter().GetResult();
            response.AddSaveChangesResult(result, "Activiteit aangepast of toegevoegd", "Activiteit niet aangepast of toegevoegd");

            response.InsertedId = entity.ActivityId;
            return response;
        }

        public Response InsertUpdateGroup(ActivityGroupBO bo)
        {
            var response = new Response();

            if (string.IsNullOrWhiteSpace(bo?.Name))
            {
                response.AddError("name is mandatory");
                return response;
            }

            ActivityGroup entity = null;

            if (bo.ID == 0)
                entity = _uow.ActivityGroups.GetNew();
            else
                entity = _uow.ActivityGroups.GetById(bo.ID);

            if (entity == null)
            {
                response.AddError("activitygroup not found");
                return response;
            }

            var err = ActivityGroupTranslator.TranslateBOToEntity(entity, bo);
            if (err != ErrorCode.Success)
            {
                response.AddError(err.ToString());
                return response;
            }

            var result = _uow.SaveChangesAsync().GetAwaiter().GetResult();
            response.AddSaveChangesResult(result, "Activiteitgroep aangepast of toegevoegd", "Activiteitgroep niet aangepast of toegevoegd");

            return response;
        }

        public Response Delete(List<int> ids)
        {
            var response = new Response();
            if (ids == null || ids.Count == 0) return response;

            foreach (var id in ids)
                _uow.Activities.DeleteObject(id);

            var result = _uow.SaveChangesAsync().GetAwaiter().GetResult();
            response.AddSaveChangesResult(result, "Record(s) verwijderd", "Geen records verwijderd");

            return response;
        }

        public Response Delete(List<ActivityBO> bos)
        {
            return Delete(bos?.Select(s => s.ID).ToList() ?? new List<int>());
        }

        public Response DeleteGroup(List<int> ids)
        {
            var response = new Response();
            if (ids == null || ids.Count == 0) return response;

            foreach (var id in ids)
                _uow.ActivityGroups.DeleteObject(id);

            var result = _uow.SaveChangesAsync().GetAwaiter().GetResult();
            response.AddSaveChangesResult(result, "Record(s) verwijderd", "Geen records verwijderd");

            return response;
        }

        public Response DeleteGroup(List<ActivityBO> bos)
        {
            // Let op: jouw vorige code riep hier Delete(bos...) aan (die Activities delete),
            // dat lijkt me niet de bedoeling voor groups. Dus we mappen naar group-ids als je die hebt.
            // Als bos eigenlijk groups bevat met ID = groupId, is dit oké:
            return DeleteGroup(bos?.Select(b => b.ID).ToList() ?? new List<int>());
        }
    }
}

File: Service/ActivityService.vb
Imports BO
Imports DAL
Imports Facade
Imports System.Threading

Public Class ActivityService
    Implements IActivityService

    Public Function GetActivitiesForSelect() As GetResponse(Of IdNameBO) Implements IActivityService.GetActivitiesForSelect
        Dim response As New GetResponse(Of IdNameBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetActivityDAO()
        Dim entities = dao.GetNoTracking()
        entities = entities.OrderBy(Function(m) m.Omschrijving)
        For Each _entity In entities
            response.AddValue(_entity.GetIdName())
        Next
        response.Values.OrderBy(Function(m) m.Display)
        Return response
    End Function

    Public Function GetActivities() As GetResponse(Of ActivityBO) Implements IActivityService.GetActivities
        Dim response As New GetResponse(Of ActivityBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetActivityDAO()
        Dim entities = dao.GetNoTracking().OrderBy(Function(m) m.Omschrijving)
        For Each _entity In entities
            Dim bo As New ActivityBO()
            bo.ID = _entity.ActivityID
            bo.Name = _entity.Omschrijving
            If (_entity.ActivityGroup IsNot Nothing) Then
                bo.Group.Name = _entity.ActivityGroup.Name
                bo.Group.ID = _entity.ActivityGroup.GroupID
                bo.Group.Lot = _entity.ActivityGroup.Lot
            End If
            'bo.GroupName = _entity.ActivityGroup.Name

            response.AddValue(bo)
        Next
        Return response
    End Function
    Public Function GetActivitiesbyId(IdList As List(Of Integer)) As GetResponse(Of ActivityBO) Implements IActivityService.GetActivitiesById
        Dim response As New GetResponse(Of ActivityBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetActivityDAO()

        Dim entities = dao.GetNoTracking.Where(Function(m) IdList.Contains(m.ActivityID))
        For Each _entity In entities
            Dim bo As New ActivityBO()
            bo.ID = _entity.ActivityID
            bo.Name = _entity.Omschrijving
            If (_entity.ActivityGroup IsNot Nothing) Then
                bo.Group.Name = _entity.ActivityGroup.Name
                bo.Group.ID = _entity.ActivityGroup.GroupID
                bo.Group.Lot = _entity.ActivityGroup.Lot
            End If
            response.AddValue(bo)
        Next
        Return response
    End Function
    Public Function GetActivitybyId(id As Integer) As GetResponse(Of ActivityBO) Implements IActivityService.GetActivitybyId
        Dim response As New GetResponse(Of ActivityBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetActivityDAO()

        Dim _entity = dao.GetById(id)
        'For Each _entity In entities
        Dim bo As New ActivityBO()
        bo.ID = _entity.ActivityID
        bo.Name = _entity.Omschrijving
        If (_entity.ActivityGroup IsNot Nothing) Then
            bo.Group.Name = _entity.ActivityGroup.Name
            bo.Group.ID = _entity.ActivityGroup.GroupID
            bo.Group.Lot = _entity.ActivityGroup.Lot
        End If
        response.AddValue(bo)
        'Next
        Return response
    End Function
    Public Function GetActivityGroups() As GetResponse(Of ActivityGroupBO) Implements IActivityService.GetActivityGroups
        Dim response As New GetResponse(Of ActivityGroupBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetActivityGroupDAO()

        Dim entities = dao.GetNoTracking()
        For Each _entity In entities
            Dim bo As New ActivityGroupBO()
            Dim err = ActivityGroupTranslator.TranslateEntityToBO(_entity, bo)
            If err = ErrorCode.Success Then
                response.AddValue(bo)
            Else
                response.AddError(err.ToString())
            End If
        Next
        Return response
    End Function
    Public Function GetActivityGroupsForSelect() As GetResponse(Of IdNameBO) Implements IActivityService.GetActivityGroupsForSelect
        Dim response As New GetResponse(Of IdNameBO)
        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetActivityGroupDAO()
        Dim entities = dao.GetNoTracking()
        entities = entities.OrderBy(Function(m) m.Lot)
        For Each _entity In entities
            response.AddValue(_entity.GetIdName())
        Next
        Return response
    End Function
    Public Function GetActivityGroupbyId(id As Integer) As GetResponse(Of ActivityGroupBO) Implements IActivityService.GetActivityGroupbyId
        Dim response As New GetResponse(Of ActivityGroupBO)

        Dim uow As New UnitOfWork(False)
        Dim dao = uow.GetActivityGroupDAO()

        Dim _entity = dao.GetById(id)
        'For Each _entity In entities
        Dim bo As New ActivityGroupBO
        bo.ID = _entity.GroupID
        bo.Name = _entity.Name
        bo.Lot = _entity.Lot
        For Each Activity In _entity.Activity
            Dim act As New ActivityBO
            act.Name = Activity.Omschrijving
            act.ID = Activity.ActivityID
            bo.Activities.Add(act)
        Next
        response.AddValue(bo)
        'Next
        Return response
    End Function


    Function InsertUpdate(bo As ActivityBO) As Response Implements IActivityService.InsertUpdate
        Dim response As New Response()
        If (String.IsNullOrWhiteSpace(bo.Name)) Then
            response.AddError("name is mandatory")
        End If
        If (Not response.Success) Then Return response

        Dim uow As New UnitOfWork()
        Dim dao = uow.GetActivityDAO()
        Dim _entity As Activity = Nothing

        If (bo.ID = 0) Then
            _entity = dao.GetNew()
        Else
            _entity = dao.GetById(bo.ID)

        End If

        If (_entity IsNot Nothing) Then

            Dim err = ActivityTranslator.TranslateBOToEntity(_entity, bo)

            If (err <> ErrorCode.Success) Then
                response.AddError(err.ToString())
            End If
        Else
            response.AddError("activity not found")
        End If
        response.AddError(uow.SaveChanges())

        Return response
    End Function
    Function InsertUpdateGroup(bo As ActivityGroupBO) As Response Implements IActivityService.InsertUpdateGroup
        Dim response As New Response()
        If (String.IsNullOrWhiteSpace(bo.Name)) Then
            response.AddError("name is mandatory")
        End If
        If (Not response.Success) Then Return response

        Dim uow As New UnitOfWork()
        Dim dao = uow.GetActivityGroupDAO()
        Dim _entity As ActivityGroup = Nothing

        If (bo.ID = 0) Then
            _entity = dao.GetNew()
        Else
            _entity = dao.GetById(bo.ID)

        End If

        If (_entity IsNot Nothing) Then

            Dim err = ActivityGroupTranslator.TranslateBOToEntity(_entity, bo)

            If (err <> ErrorCode.Success) Then
                response.AddError(err.ToString())
            End If
        Else
            response.AddError("activity not found")
        End If
        response.AddError(uow.SaveChanges())

        Return response
    End Function

    Function Delete(ids As List(Of Integer)) As Response Implements IActivityService.Delete
        Dim response As New Response()
        Dim uow As New UnitOfWork()

        For Each id In ids
            uow.GetActivityDAO().DeleteObject(id)
        Next
        response.Messages.AddRange(uow.SaveChanges())

        Return response
    End Function

    Function Delete(bos As List(Of ActivityBO)) As Response Implements IActivityService.Delete
        Return Delete(bos.Select(Function(s) s.ID).ToList())
    End Function
    Function DeleteGroup(ids As List(Of Integer)) As Response Implements IActivityService.DeleteGroup
        Dim response As New Response()
        Dim uow As New UnitOfWork()

        For Each id In ids
            uow.GetActivityGroupDAO().DeleteObject(id)
        Next
        response.Messages.AddRange(uow.SaveChanges())

        Return response
    End Function

    Function DeleteGroup(bos As List(Of ActivityBO)) As Response Implements IActivityService.DeleteGroup
        Return Delete(bos.Select(Function(s) s.ID).ToList())
    End Function
End Class
